igvm/commands.py

```diff
@@ -569,22 +616,49 @@ def _get_best_hypervisor(vm, hypervisor_states, offline=False):
         'state': Any(*hypervisor_states),
     }, HYPERVISOR_ATTRIBUTES))
 
+    ret_hypervisor = None
     for hypervisor in sorted_hypervisors(
         HYPERVISOR_PREFERENCES, vm, hypervisors
     ):
         # The actual resources are not checked during sorting for performance.
         # We need to validate the hypervisor using the actual values before
         # the final decision.
+        try:
+            hypervisor.lock()
+        except InvalidStateError as error:
+            log.warning(error)
+            continue
+
         try:
             hypervisor.check_vm(vm, offline)
         except HypervisorError as error:
+            hypervisor.unlock()
             log.warning(
-                'Preferred hypervisor "{}" is skipped:  {}'
+                'Preferred hypervisor "{}" is skipped: {}'
                 .format(hypervisor, error)
             )
-        else:
-            return hypervisor
-    raise IGVMError('Cannot find a hypervisor')
+            continue
+
+        ret_hypervisor = hypervisor
+        break
+
+    if ret_hypervisor:
```

Comment:
I don't see why you added `ret_hypervisor`.  You could just `yield` inside the loop and `break`.  The indentation would be the same.  In short, you can just remove lines 642-645.

Reply:
I did not think of yielding from loop. Good idea.

Reply:
`yield`+`break` from inside of loop breaks handling exceptions. I don't really understand how it happens.